skimage/restoration/inpaint.py

```diff
@@ -1,79 +1,213 @@
 
-    # Find masked points and prepare them to be easily enumerate over
-    mask_pts = np.stack(np.where(mask), axis=-1)
+    This function creates a linear system of the form:
+
+    ``A @ u = b``
+
+    where ``A`` is a sparse matrix, ``b`` is a vector enforcing smoothness and
+    boundary constraints and ``u`` is the vector of inpainted values to be
+    (uniquely) determined by solving the linear system.
+
+    ``A`` is a sparse matrix of shape (n_mask, n_mask) where `n_mask``
+    corresponds to the number of non-zero values in ``mask`` (i.e. the number
+    of pixels to be inpainted). Each row in A will have a number of non-zero
+    values equal to the number of non-zero values in the biharmonic kernel,
+    ``neigh_coef_full``. In practice, biharmonic kernels with reduced extent
+    are used at the image borders. This matrix, ``A`` is the same for all
+    image channels (since the same inpainting mask is currently used for all
+    channels).
+
+    ``u`` is a dense matrix of shape ``(n_mask, n_channels)`` and represents
+    the vector of unknown values for each channel.
+
+    ``b`` is a dense matrix of shape ``(n_mask, n_channels)`` and represents
+    the desired output of convolving the solution with the biharmonic kernel.
+    At mask locations where there is no overlap with known values, ``b`` will
+    have a value of 0. This enforces the biharmonic smoothness constraint in
+    the interior of inpainting regions. For regions near the boundary that
+    overlap with known values, the entries in ``b`` enforce boundary conditions
+    designed to avoid discontinuity with the known values.
+    """
 
-    # Iterate over masked points
-    for mask_pt_n, mask_pt_idx in enumerate(mask_pts):
+    n_channels = out.shape[-1]
+    radius = neigh_coef_full.shape[0] // 2
+
+    edge_mask = np.ones(mask.shape, dtype=bool)
+    edge_mask[(slice(2, -2),) * mask.ndim] = 0
+    boundary_mask = edge_mask * mask
+    center_mask = ~edge_mask * mask
+
+    boundary_pts = np.where(boundary_mask)
+    boundary_i = np.where(boundary_mask.ravel())[0]
+    center_i = np.where(center_mask.ravel())[0]
+    mask_i = np.concatenate((boundary_i, center_i))
+
+    center_pts = np.where(center_mask)
+    mask_pts = tuple(
+        [np.concatenate((b, c)) for b, c in zip(boundary_pts, center_pts)]
+    )
+
+    # Use convolution to predetermine the number of non-zero entries in the
+    # sparse system matrix.
+    structure = neigh_coef_full != 0
+    tmp = ndi.convolve(mask, structure, output=np.uint8, mode='constant')
+    nnz_matrix = tmp[mask].sum()
+
+    # Need to estimate the number of zeros for the right hand side vector.
+    # The computation below will slightly overestimate the true number of zeros
+    # due to edge effects (the kernel itself gets shrunk in size near the
+    # edges, but that isn't accounted for here). We can trim any excess entries
+    # later.
+    n_mask = mask.sum()
```

Comment:
`np.count_nonzero` is much faster then `sum`:

```python
In [1]: x = np.random.rand(500, 500) > 0.5

In [2]: %timeit np.count_nonzero(x)                                                                     
9.96 µs ± 71.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [3]: %timeit x.sum()                                                                                 
196 µs ± 150 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
```suggestion
    n_mask = np.count_nonzero(mask)
```
:wink: